src/api/useFetch.js

```diff
@@ -17,6 +17,210 @@ import {
+ * @param {Object} fetchParams - parameters to pass to the fetch function, provide `null` to not pass
+ * parameters at all.
+ * @param {Function} [fetchCondition=null] - optional callback which shortcuts fetching if it is falsy.
+ * @param {string} [errorMessage='error'] - error message to display in the snackbar.
+ * @param {Function} [retryCondition] - used to determine if a retry is needed based on theresult.
+ * @param {number} [retryIntervalSeconds=10] - interval in seconds between retry attempts.
+ * @param {boolean} [showSnackbar=true] - flag to show a snackbar on error.
+ * @param {string} [cacheKey=JSON.stringify({fetchFunction, fetchParams})] - if not `null`, use to
+ * lookup and serve result from cache.
+ * @param {Function} [cacheCondition=null] - optional callback, takes resolved promise data as input
+ * and must be truthy to write entry to cache (implication: prevents unresolved promise from being
+ * cached)
+ * @returns {[any, {isLoading: boolean, error: string|null}, callable]} - array containing the
+ * fetched data, the state object, and a callback to re-fetch (data state is *not* re-initialised
+ * allowing consumers to differeniate between inital fetch and re-fetch).
+ */
+const _useFetch = ({
+  fetchFunction,
+  fetchParams = null,
+  fetchCondition = null,
+  errorMessage = 'error',
+  retryCondition,
+  retryIntervalSeconds = 10,
+  showSnackbar = true,
+  cacheCondition = null,
+  cacheKey = JSON.stringify({
+    fetchFunction: fetchFunction.name,
+    fetchParams: fetchParams === null ? null : normaliseObject(fetchParams),
+  }),
+}) => {
+  const [data, setData] = React.useState()
+  const [state, setState] = React.useState({
+    isLoading: true,
+    error: null,
+  })
+
+  // toggle state to refresh useFetch hook, callable is returned next to data and state
+  const [refresh, setRefresh] = React.useState(true)
+  const resetState = () => {
+    setState({
+      isLoading: true,
+      error: null,
+    })
+  }
+  React.useEffect(() => resetState(), [refresh])
+
+  const retryIntervalRef = React.useRef()
+  const isFetching = React.useRef(false)
+
+  const { enqueueSnackbar } = useSnackbar()
+
+  const clearIntervalFromRef = (ref) => {
+    clearInterval(ref.current)
+    ref.current = null
+  }
+
+  React.useEffect(() => {
+    if (fetchCondition && !fetchCondition()) return
+
+    // track whether hook is mounted
+    let mounted = true
+
+    const fetchData = async () => {
+      // prevent multiple fetch attempts if component is re-rendered while fetching
+      if (isFetching.current) return
+      isFetching.current = true
+
+      // prevent stale updates to avoid memory leaks
+      if (!mounted) return
+
+      if (cacheKey !== null) {
+        const cachedResult = cache.get(cacheKey)
+        if (cachedResult) {
+          if (cachedResult instanceof Promise) {
+            try {
+              const result = await cachedResult
```

Comment:
Shouldn't we also write the actual result in the cache once we successfully retrieved it?

Reply:
Cache is updated in line 167. However, as promises can be resolved multiple times, this wouldn't even be necessary (even though it is of course reasonable considering performance).

Reply:
Yes but it will never be reached in case a promise has been cached because of the `return`, will it?

Reply:
Hm, you might be correct now that I think about the retry case. Will reconsider tomorrow once I have had more coffee x)

Reply:
You are right, in retry case the cached entry was not updated from promise to resolved promise. While debugging I also discovered a second flaw, the cache was also not updated if not `cacheCondition` was provided.

Fixed both, good catch!